### PR TITLE
Add prefer-align-cpus-by-uncorecache feature gate

### PR DESCRIPTION
**Pull request by wongchar**
### Description

Add documentation for new alpha feature gate: prefer-align-cpus-by-uncorecache

### Issue

https://github.com/kubernetes/enhancements/issues/4800
https://github.com/kubernetes/kubernetes/pull/126750

### CODE REVIEW

**Comment by tengqm** [comment]
This commit should be rebased onto `dev-1.32` head, IIUC.

**Comment by tengqm** [comment]
@wongchar Please rebase your PR to the `dev-1.32` HEAD. Your current PR is difficult to review. It contains 44 commits to 37 files.

**Comment by chanieljdan** [comment]
Hi @wongchar 👋 

Just a reminder to take a look at [Documenting for a release - PR Ready for Review](https://kubernetes.io/docs/contribute/new-content/new-features/#pr-ready-for-review) to get your PR ready for review ahead of the "Ready for Review" deadline on Tuesday November 19th 2024 18:00 PST. 

Thank you!

**Comment by chanieljdan** [comment]
Hi @wongchar 👋,

A quick reminder to get your PR ready for review before the deadline, Thursday Oct 24 at 18:00 PDT. For additional information, refer to: [Documenting for a release - PR Ready for Review](https://kubernetes.io/docs/contribute/new-content/new-features/#pr-ready-for-review).

Thank you!

File: content/en/docs/tasks/administer-cluster/cpu-management-policies.md [hunk 1]
```diff
@@ -318,6 +319,15 @@ the benefit of reducing contention diminishes. Conversely, default behavior
 can help in reducing inter-core communication overhead, potentially providing
 better performance under high load conditions.
 
+If the `prefer-align-cpus-by-uncorecache` policy is specified, the static policy
+will allocate CPU resources for individual containers by uncore cache. By default,
```
Comment (ffromani) [inline content/en/docs/tasks/administer-cluster/cpu-management-policies.md:323]: suggestion: 
```
If the `prefer-align-cpus-by-uncorecache` policy is specified, the static policy will allocate CPU resources for individual containers aligning at the uncore (sometimes known as L3) boundary. This means the CPUs allocated to the container will all share the same uncore cache block.
```

File: content/en/docs/tasks/administer-cluster/cpu-management-policies.md [hunk 2]
```diff
@@ -318,6 +319,15 @@ the benefit of reducing contention diminishes. Conversely, default behavior
 can help in reducing inter-core communication overhead, potentially providing
 better performance under high load conditions.
 
+If the `prefer-align-cpus-by-uncorecache` policy is specified, the static policy
+will allocate CPU resources for individual containers by uncore cache. By default, 
+the `CPUManager` will tightly pack CPU assignments which can result in containers 
+being assigned CPUs from multiple uncore caches. This feature attempts to align 
+containers to CPUs that correspond with the fewest amount of uncore caches at a
+best-effort policy. Specific workloads can benefit in performance from the reduction
```
Comment (ffromani) [inline content/en/docs/tasks/administer-cluster/cpu-management-policies.md:327]: suggestion:
```
This feature attempts to align containers to CPUs that correspond with the fewest amount of uncore caches at a
best-effort policy. 
```
becomes
```
This option makes the CPU manager try to allocate cpus in such a way to make the most efficient possible use of uncore cache. The allocation is done in a best-effort fashion, in order to allocate most CPUs affine to the same uncore cache block. If more blocks are required to satisfy the allocation, the fewer amount of blocks is allocated, to maximize the uncore cache alignment.
```

File: content/en/docs/tasks/administer-cluster/cpu-management-policies.md [hunk 3]
```diff
@@ -334,3 +344,7 @@ The `distribute-cpus-across-cores` option can be enabled by adding
 `distribute-cpus-across-cores=true` to the `CPUManager` policy options.
 It cannot be used with `full-pcpus-only` or `distribute-cpus-across-numa` policy
 options together at this moment.
+
+The `prefer-align-cpus-by-uncorecache` option can be enabled by adding the
+`prefer-align-cpus-by-uncorecache` to the `CPUManager` policy options. It cannot
+be used with `distribute-cpus-across-numa` or `distribute-cpus-across-cores`
```
Comment (ffromani) [inline content/en/docs/tasks/administer-cluster/cpu-management-policies.md:350]: suggestion:
```
if incompatible options are used, the kubelet will fail to start with the error explained in the logs.
```

**Review by ffromani — COMMENTED** [review]
the content is technically correct but we should give a bit more context to the users.
The main issue I see myself is that users which are not already aware of how LLC alignment can help them don't gain much knowledge from the added docs. We should try to guide them a bit more, still considering this page is not the best place to introduce LLC concepts.

File: content/en/docs/tasks/administer-cluster/cpu-management-policies.md (hunk 3 above)
Comment (wongchar) [reply content/en/docs/tasks/administer-cluster/cpu-management-policies.md]: Thanks @ffromani 
Applied the suggested changes, please take a look when you get the chance, thanks!

File: content/en/docs/reference/command-line-tools-reference/feature-gates/prefer-align-cpus-by-uncorecache.md [hunk 4]
```diff
@@ -0,0 +1,13 @@
...
+_build:
+  list: never
+  render: false
+
+stages:
+  - stage: alpha
+    defaultValue: false
+    fromVersion: "1.32"
+---
+When `PreferAlignCpusByUncoreCache` is enabled while the CPU Manager Policy is set to `static`, containers within a `Guaranteed` pod will individually be aligned to an uncore cache group at a best-effort policy. This feature can optimize performance for certain cache-sensitive workloads by minimizing the cpu allocation across uncore caches.
```
Comment (sftim) [inline content/en/docs/reference/command-line-tools-reference/feature-gates/prefer-align-cpus-by-uncorecache.md:13]: Try this
```suggestion
Enable support for aligning workloads to a cache group that spans multiple cores: an _uncore_ cache group.

To learn more, read [CPU management policies](/docs/tasks/administer-cluster/cpu-management-policies/).
```

**Review by sftim — COMMENTED** [review]
We have reached a point where the existing task page is quite hard to follow.

To be able to document this well, ideally we'd have a **concept** page about CPU management policies.

See https://kubernetes.io/docs/concepts/policy/ - can you add a very brief page into that section that explains the idea?

Then, the feature gate explanation can include a hyperlink to the concept page.

---

If you have the appetite to do this well (which needs more effort), consider moving the list of available policies to live within https://kubernetes.io/docs/reference/node/

File: content/en/docs/reference/command-line-tools-reference/feature-gates/prefer-align-cpus-by-uncorecache.md (hunk 4 above)
Comment (sftim) [reply content/en/docs/reference/command-line-tools-reference/feature-gates/prefer-align-cpus-by-uncorecache.md:13]: This should ideally link to a concept page instead; see https://github.com/kubernetes/website/pull/48340#pullrequestreview-2443697490

Comment (wongchar) [reply content/en/docs/reference/command-line-tools-reference/feature-gates/prefer-align-cpus-by-uncorecache.md:13]: > See https://kubernetes.io/docs/concepts/policy/ - can you add a very brief page into that section that explains the idea?
This feature is a CPU Manager Policy (under Node Resource Managers). Just want to confirm my understanding that you are asking for a standalone concept page for `prefer-align-cpus-by-uncorecache` that is separate from the existing CPU Manager Policies concept page?

**Comment by ffromani** [comment]
Hi @sftim ! My understanding is we have time to the doc freeze and Nov 19 was the deadline to have documentation PRs *submitted* and ready for review (not merged!)

If my understanding is right I'm willing to help to kickoff the aforementioned concepts page.

**Comment by sftim** [comment]
@ffromani that's right. Help is welcome!

**Comment by ffromani** [comment]
> @ffromani that's right. Help is welcome!

ok! I'm willing to take a stab! (xref: https://github.com/kubernetes/website/issues/38121)

**Comment by ffromani** [comment]
> > @ffromani that's right. Help is welcome!
> 
> ok! I'm willing to take a stab! (xref: #38121)

here: https://github.com/kubernetes/website/pull/48797

**Review by ffromani — COMMENTED** [review]
LGTM from sig-node. Technical content looks correct to me.

**Comment by chanieljdan** [comment]
Technical lgtm provided above

/lgtm

**Comment by chanieljdan** [comment]
@reylejano when you get the chance we just need your final review and approval for this one. Thanks!

File: content/en/docs/tasks/administer-cluster/cpu-management-policies.md
Comment (sftim) [inline content/en/docs/tasks/administer-cluster/cpu-management-policies.md:1]: If we can get https://github.com/kubernetes/website/pull/48797 merged and merged into v1.32 docs, then the changes here can go there instead (which is what I recommend doing).

See https://github.com/kubernetes/website/pull/48340#pullrequestreview-2443697490

File: content/en/docs/reference/command-line-tools-reference/feature-gates/prefer-align-cpus-by-uncorecache.md (hunk 4 above)
Comment (sftim) [reply content/en/docs/reference/command-line-tools-reference/feature-gates/prefer-align-cpus-by-uncorecache.md:13]: > Just want to confirm my understanding that you are asking for a standalone concept page for `prefer-align-cpus-by-uncorecache` that is separate from the existing CPU Manager Policies concept page?


Not that, no.

I'm asking for this feature gate description to hyperlink to a concept page about CPU manager policies. The target of the hyperlink doesn't have to be specifically about `prefer-align-cpus-by-uncorecache`.

Comment (sftim) [reply content/en/docs/reference/command-line-tools-reference/feature-gates/prefer-align-cpus-by-uncorecache.md:13]: The https://kubernetes.io/docs/concepts/policy/node-resource-managers/ concept page is _too basic_ and SIG Node should put some help towards making it more reader friendly. SIG Docs will help.

File: content/en/docs/tasks/administer-cluster/cpu-management-policies.md
Comment (ffromani) [reply content/en/docs/tasks/administer-cluster/cpu-management-policies.md:1]: I'm willing to help with #48797, but @tengqm raised some points about the feasibility in the 1.32 timeframe. I guess we can figure out during the review of #48797 ?

**Comment by sftim** [comment]
Waiting for docs LGTM
Tech LGTMs noted

/lgtm cancel

File: content/en/docs/tasks/administer-cluster/cpu-management-policies.md
Comment (ffromani) [reply content/en/docs/tasks/administer-cluster/cpu-management-policies.md:1]: I can and will work on getting #48797 ready anyway (https://kubernetes.slack.com/archives/C0BP8PW9G/p1732284959539839). If we decide to not merge for 1.32 due to the too strict timeline, I'll rebase once the other linked content landed, and once cpumanager content is settled I'll move on to memory manager and others:
1. cpumanager
2. memory manager
3. topology manager
4. (perhaps) device manager, this component doesn't have policies so it can use some refactoring but probably doesn't belong here.

**Comment by wongchar** [comment]
Rebased to @ffromani 's PR (https://github.com/kubernetes/website/pull/48856). Thanks!

@sftim PTAL when you get a chance.

File: content/en/docs/concepts/policy/node-resource-managers.md [hunk 5]
```diff
@@ -189,11 +189,15 @@ The following policy options exist for the static CPU management policy:
 `align-by-socket` (alpha, hidden by default)
 : Align CPUs by physical package / socket boundary, rather than logical NUMA boundaries (available since Kubernetes v1.25)
 `distribute-cpus-across-cores` (alpha, hidden by default)
-: allocate virtual cores, sometimes called hardware threads, across different physical cores  (available since Kubernetes v1.31)
+: Allocate virtual cores, sometimes called hardware threads, across different physical cores  (available since Kubernetes v1.31)
 `distribute-cpus-across-numa` (alpha, hidden by default)
-: spread CPUs across different NUMA domains, aiming for an even balance between the selected domains (available since Kubernetes v1.23)
+: Spread CPUs across different NUMA domains, aiming for an even balance between the selected domains (available since Kubernetes v1.23)
```
Comment (ffromani) [inline content/en/docs/concepts/policy/node-resource-managers.md:194]: this is unrelated fix I asked @wongchar to pull in

File: content/en/docs/concepts/policy/node-resource-managers.md [hunk 6]
```diff
@@ -189,11 +189,15 @@ The following policy options exist for the static CPU management policy:
...
 `full-pcpus-only` (beta, visible by default)
 : Always allocate full physical cores (available since Kubernetes v1.22)
+`strict-cpu-reservation` (alpha, hidden by default)
+: Prevent all the pods regardless of their Quality of Service class to run on reserved CPUs (available since Kubernetes v1.32)
```
Comment (ffromani) [inline content/en/docs/concepts/policy/node-resource-managers.md:198]: Same, fixing a rebase artifact of my previous https://github.com/kubernetes/website/pull/48797

File: content/en/docs/reference/command-line-tools-reference/feature-gates/prefer-align-cpus-by-uncorecache.md (hunk 4 above)
Comment (ffromani) [reply content/en/docs/reference/command-line-tools-reference/feature-gates/prefer-align-cpus-by-uncorecache.md:13]: fixed in the last upload after rebase on top of https://github.com/kubernetes/website/pull/48797

**Review by ffromani — COMMENTED** [review]
LGTM

from sig-node

**Comment by chanieljdan** [comment]
tech and docs lgtm provided

/approve
